xc/device.ml

```diff
@@ -1828,6 +1857,8 @@ let stop ~xs ~qemu_domid domid  =
 		| Some qemu_pid ->
 			debug "qemu-dm: stopping qemu-dm with SIGTERM (domid = %d)" domid;
 			let open Generic in
+			    best_effort "qemu did not crash"
```

Comment:
The qemu-pid-signal key is set at most once, in one of 2 different situations:

- inside this function (device.stop), which is an expected shutdown of qemu. In this case, we indicate explicitly that we want no processing of any signals by setting an empty value for this key.

- outside this function, indicating an unexpected shutdown of qemu. In this case, we set the key value to be the signal that arrived from qemu as a hint for further processing by xenopsd to take further action. In this patch, this set operation is guarded by a check in waitpid_async to verify if the key is already present. If it is, then it's not set a second time -- we don't expect a 2nd signal to arrive from the same qemu pid, and once this value is set, it should not change. (A nice effect of this guard is that after an expected shutdown of qemu in device.stop, where we set the value to be empty, then we'll not receive spurious signals from forkexecd for which we are not interested anymore.)